apollo-federation/src/query_plan/optimize.rs

```diff
@@ -0,0 +1,755 @@
+    ) {
+        // Note: It's not possible for two fragments to include each other. So, we don't need to
+        //       worry about inclusion cycles.
+        let included_fragments: HashSet<Name> = applicable_fragments
+            .iter()
+            .filter(|(fragment, _)| {
+                applicable_fragments
+                    .iter()
+                    .any(|(other_fragment, _)| other_fragment.includes(&fragment.name))
+            })
+            .map(|(fragment, _)| fragment.name.clone())
+            .collect();
+
+        applicable_fragments.retain(|(fragment, _)| !included_fragments.contains(&fragment.name));
+    }
+
+    /// Try to optimize the selection set by re-using existing fragments.
+    /// Returns either
+    /// - a new selection set partially optimized by re-using given `fragments`, or
+    /// - a single fragment that covers the full selection set.
+    // PORT_NOTE: Moved from `Selection` class in JS code to SelectionSet struct in Rust.
+    // PORT_NOTE: `parent_type` argument seems always to be the same as `self.type_position`.
+    fn try_optimize_with_fragments(
+        &self,
+        parent_type: &CompositeTypeDefinitionPosition,
+        fragments: &NamedFragments,
+        validator: &mut FieldsConflictMultiBranchValidator,
+        can_use_full_matching_fragment: impl Fn(&Fragment) -> bool,
+    ) -> Result<SelectionSetOrFragment, FederationError> {
+        // We limit to fragments whose selection could be applied "directly" at `parent_type`,
+        // meaning without taking the fragment condition into account. The idea being that if the
+        // fragment condition would be needed inside `parent_type`, then that condition will not
+        // have been "normalized away" and so we want for this very call to be called on the
+        // fragment whose type _is_ the fragment condition (at which point, this
+        // `can_apply_directly_at_type` method will apply. Also note that this is because we have
+        // this restriction that calling `expanded_selection_set_at_type` is ok.
+        let candidates = fragments.get_all_may_apply_directly_at_type(parent_type)?;
+        if candidates.is_empty() {
+            return Ok(self.clone().into()); // Not optimizable
+        }
+
+        // First, we check which of the candidates do apply inside the selection set, if any. If we
+        // find a candidate that applies to the whole selection set, then we stop and only return
+        // that one candidate. Otherwise, we cumulate in `applicable_fragments` the list of fragments
+        // that applies to a subset.
+        let mut applicable_fragments = Vec::new();
+        for candidate in candidates {
+            let at_type = candidate.expanded_selection_set_at_type(parent_type)?;
+            if at_type.is_useless() {
+                continue;
+            }
+            if !validator.check_can_reuse_fragment_and_track_it(&at_type)? {
+                continue;
```

Comment:
for posterity's sake
```suggestion
                // We cannot use it at all, so no point in adding to `applicable_fragments`.
                continue;
```